src/harness/core.ts

```diff
@@ -0,0 +1,671 @@
+            }
+        }
+
+        public [Symbol.iterator]() {
+            return this.values();
+        }
+
+        private writePreamble() {
+            if (this._copyOnWrite) {
+                this._values = this._values.slice();
+                this._copyOnWrite = false;
+            }
+        }
+
+        private writePostScript() {
+            this._version++;
+        }
+    }
+
+    /**
+     * A collection of metadata that supports inheritance.
+     */
+    export class Metadata {
+        private static readonly _undefinedValue = {};
+        private _parent: Metadata | undefined;
+        private _map: { [key: string]: any };
+        private _version = 0;
+        private _size = -1;
+        private _parentVersion: number | undefined;
+
+        constructor(parent?: Metadata) {
+            this._parent = parent;
+            this._map = Object.create(parent ? parent._map : null); // tslint:disable-line:no-null-keyword
+        }
+
+        public get size(): number {
+            if (this._size === -1 || (this._parent && this._parent._version !== this._parentVersion)) {
+                let size = 0;
+                for (const _ in this._map) size++;
+                this._size = size;
+                if (this._parent) {
+                    this._parentVersion = this._parent._version;
+                }
+            }
+            return this._size;
+        }
+
+        public has(key: string): boolean {
+            return this._map[Metadata._escapeKey(key)] !== undefined;
+        }
+
+        public get(key: string): any {
+            const value = this._map[Metadata._escapeKey(key)];
+            return value === Metadata._undefinedValue ? undefined : value;
+        }
+
+        public set(key: string, value: any): this {
+            this._map[Metadata._escapeKey(key)] = value === undefined ? Metadata._undefinedValue : value;
+            this._size = -1;
+            this._version++;
+            return this;
+        }
+
+        public delete(key: string): boolean {
+            const escapedKey = Metadata._escapeKey(key);
+            if (this._map[escapedKey] !== undefined) {
+                delete this._map[escapedKey];
+                this._size = -1;
+                this._version++;
+                return true;
+            }
+            return false;
+        }
+
+        public clear(): void {
+            this._map = Object.create(this._parent ? this._parent._map : null); // tslint:disable-line:no-null-keyword
+            this._size = -1;
+            this._version++;
+        }
+
+        public forEach(callback: (value: any, key: string, map: this) => void) {
+            for (const key in this._map) {
+                callback(this._map[key], Metadata._unescapeKey(key), this);
+            }
+        }
+
+        private static _escapeKey(text: string) {
+            return (text.length >= 2 && text.charAt(0) === "_" && text.charAt(1) === "_" ? "_" + text : text);

```

Comment:
I feel like these should be using the `escapeLeadingUnderscores` and `unescapeLeadingUnderscores` methods in `utilities` to avoid duplicating functionality.

Reply:
See comment at top of file.